### ISSUE TITLE
Inline styles

### ISSUE THREAD

**Issue by jkcclemens** [kazzkiq/CodeFlask#64]
Firstly, looks like you released 1.0! Congrats! :D

Unfortunately, it seems that you have introduced inline styles with this update. As someone enforcing a strict Content Security Policy, the reason I was using CodeFlask is precisely because it *didn't* do this. Please consider changing this new update to remove the inline styles and be more CSP-friendly!

**Comment by lol768** [comment]
:+1: There are far too many JS libraries which do this and negatively impact oberall security

**Comment by jkcclemens** [comment]
Is it possible to replace [this](https://github.com/kazzkiq/CodeFlask/blob/e64a3082a3ce4ce8ad9db60ad9e600f3d2944a16/src/styles/editor.js) with a Sass workflow? One base theme that can be customized by variables?

```scss
@import 'variables';

$token: red;

@import 'base';
```

**Comment by kazzkiq** [comment]
Not every user uses SASS. CodeFlask must be as much technology "agnostic" as possible. So that's out of question.

About the CSP concern, I honestly cannot see where could lie the issue with inline CSS, since:

1. It only runs only on modern browsers (which have no more issues executing crazy HTC stuff like IE had);
2. The CSS is not user-based. It is (and always will be) the exact snippet you're seeing in source-code;
3. Most (if not all) major reactive frameworks/libs use inline styles and/or CSS injection via JavaScript, and that doesn't seem like a big deal.

And of course, the main goal by injecting CSS directly via JavaScript is to make it even easier to get CodeFlask up and running. Just import it, run on some element, and you have an editor with basic highlight support.

I'm honestly interested in the threats such approach could lead to. If that's a big concern, what could be done is adding an option to prevent any CSS injection, so its up to users if they want to add all the CSS by themselves or let CodeFlask handle it.

**Comment by jkcclemens** [comment]
What I meant is that *you* use Sass to generate the static CSS.

I'm sure @lol768 could explain more in depth about why inline CSS is unsafe (the CSP keyword is literally `unsafe-inline`), but OWASP has some good examples of how to use injected CSS to facilitate XSS attacks.

Having to make an exception and allow ***all*** unsafe inline styles to use CodeFlask means that I won't be using it. There's no reason to inject CSS when the `link` tag exists.

**Comment by lol768** [comment]
Hey @kazzkiq,

I personally don't care too much how the CSS is generated, I'm guessing @jkcclemens saw the interpolation and figured SASS was a good fit at built time. This is an implementation detail though, so do it however you like.

All I ask is that you please allow site owners to include the styles themselves in an external file instead of injecting it like you are now which isn't a great practice from a security perspective.

I'm not sure I understand your comments concerning modern browsers. The inline CSS (which is effectively what your style injection is) is an issue in all browsers which support CSP.

If it's static then there should be no problem with it being hosted in a separate file and having the user include it with a `<link>`.

Yeah, a lot of libraries do it and it's a major headache for someone who works in infosec. We try and recommend the strictest possible CSP but can't when the client relies on libraries which do this because it'd cause breakage.

I can totally see why you do it. It's easier for the end user, but I mean it's only an extra `<link>` or adding some CSS to your existing stylesheet..

https://github.com/ChALkeR/notes/blob/master/Improper-markup-sanitization.md is an interesting write-up. The Vanilla issue on particular is a good example of why disallowing `unsafe-inline` is a good idea. Some of the other reports use classes which already exist (and a restrictive CSP wouldn't help here) but nonetheless you can see it's useful to have.

Take a look at this one, too: http://scarybeastsecurity.blogspot.com/2009/12/generic-cross-browser-cross-domain.html

**Comment by kazzkiq** [comment]
@lol768 Thanks for your time for a more detailed explanation. I do agree with most of your points, and those articles clarify a lot, too.

I still think the best approach here would be to give user the choice of trading ease of use for CSP comply. In this case, I believe the change could go as follows:

1. Adding `csp: true|false` option (defaults to `false`);
2. Explaining in the docs the impact of either value chosen (how it affects security if `false`, and how to adapt CodeFlask styles in case of `true`);
3. Generate CSS files do `/build` in case the choice was `true`.
3. Explaining why it is a thing and why exactly the option was added to the lib (perhaps with a tiny intro to CSP and its importance).

This way I think we can satisfy both worlds. Those who want to use CodeFlask "just for that simple weekend project" can stick with the way it works out-of-the-box now. And those who are building serious stuff and care about security can also easily adapt the library to their concerns.

**Comment by lol768** [comment]
Sounds good to me :slightly_smiling:

**Comment by kazzkiq** [comment]
Great. I will implement it as soon as possible.

**Comment by myfreeweb** [comment]
Thanks for not removing inline, it's good for #68.

Note for everyone: you don't have to use `unsafe-inline`. CSP level 2 [allows allowing (heh) specific inline styles/scripts via SHA hashes or nonces](https://developers.google.com/web/fundamentals/security/csp/#if_you_absolutely_must_use_it)!

**Comment by lol768** [comment]
>Note for everyone: you don't have to use unsafe-inline. CSP level 2 allows allowing (heh) specific inline styles/scripts via SHA hashes or nonces!

Yeah, and it's undesirable. Should only be used for transitioning legacy code, will leave the page broken in CSPv1 browsers and it's painful to maintain. Better to ask the library maintainers to do things properly.

Nonces wouldn't be suitable here in any case given the current implementation since the JS doesn't introduce any nonce attributes and presumably wouldn't know what the nonce was either.

**Comment by ultimate-tester** [comment]
Did this already get implemented? I would love to make use of it, just because injecting CSS with JS is messy business
